boxoffice/models/line_item.py

```diff
@@ -1,47 +1,212 @@
+from coaster.utils import LabeledEnum
+from baseframe import __
 
-from flask import url_for
+__all__ = ['LineItem']
 
-from boxoffice.models import db, BaseMixin
-from boxoffice.models.category import Category
-from boxoffice.models.order import Order
 
-__all__ = ['LineItem']
+class LINE_ITEM_STATUS(LabeledEnum):
+    CONFIRMED = (0, __("Confirmed"))
+    CANCELLED = (1, __("Cancelled"))
 
 
 class LineItem(BaseMixin, db.Model):
     """
-    A LineItem is part of an invoice.
-    Each line item carries an item category, quantity, charged amount, tax rate & line total.
+    Note: Line Items MUST NOT be deleted.
+    They must only be cancelled.
     """
-    __tablename__ = 'lineitem'
-    order_id = db.Column(db.Integer, db.ForeignKey('order.id'), nullable=False)
-    order = db.relation(Order, primaryjoin=order_id == Order.id,
-        backref=db.backref('lineitems', cascade='all, delete-orphan'))
-
-    #: Item category of this LineItem
-    category_id = db.Column(db.Integer, db.ForeignKey('category.id'), nullable=False)
-    category = db.relationship(Category, primaryjoin=category_id == Category.id)
-
-    quantity = db.Column(db.Integer, default=1, nullable=False)
-
-    total = db.Column(db.Numeric(10, 2), nullable=True)
-
-    def update_total(self):
-        self.total = self.quantity * self.category.price_after_tax
-
-    def dict(self):
-        return self.__dict__
-
-    def as_json(self):
-        return {
-            'event': self.category.event.title,
-            'category_id': self.category_id,
-            'category_desc': self.category.title,
-            'quantity': self.quantity,
-            'pat': self.category.price_after_tax,
-            'tax_rate': self.category.tax_rate,
-            'created_at': self.created_at.isoformat() + 'Z',
-            'updated_at': self.updated_at.isoformat() + 'Z',
-        }
+    __tablename__ = 'line_item'
+    __uuid_primary_key__ = True
+    __tableargs__ = (db.UniqueConstraint('order_id', 'line_item_no'),)
+
+    customer_order_id = db.Column(None, db.ForeignKey('customer_order.id'), nullable=False)
+    order = db.relationship(Order, backref=db.backref('line_items', cascade='all, delete-orphan'))
+    # line_item_no is the relative number of the line item per order.
+    line_item_no = db.Column(db.Integer, nullable=False)
+
+    item_id = db.Column(None, db.ForeignKey('item.id'), nullable=False)
+    item = db.relationship(Item, backref=db.backref('line_items', cascade='all, delete-orphan'))
+
+    discount_policy_id = db.Column(None, db.ForeignKey('discount_policy.id'), nullable=True)
+    discount_policy = db.relationship('DiscountPolicy', backref=db.backref('line_items'))
+
+    discount_coupon_id = db.Column(None, db.ForeignKey('discount_coupon.id'), nullable=True)
+    discount_coupon = db.relationship('DiscountCoupon', backref=db.backref('line_items'))
+
+    base_amount = db.Column(db.Numeric, default=decimal.Decimal(0), nullable=False)
```

Comment:
Just to make this less verbose, use `from decimal import Decimal` above. Also, [SQLAlchemy recommends using `cdecimal` instead of `decimal`](http://docs.sqlalchemy.org/en/latest/core/type_basics.html#sqlalchemy.types.Numeric), but you'll have to patch that in elsewhere.